### ISSUE TITLE
Discover LMod modules don't work for Python

### ISSUE THREAD

**Issue by markjolah** [JCSDA/jedi-stack#95]
I think there is a bug in the version of LMod installed on discover `7.5.17`.  I am using the LMod Python interface in JEDI-Rapids, however it gets tripped up by this code which is in the top-level App provided by JEDI-Stack. `modulefiles/apps/jedi/intel17-impi`

```
# Set compilers to use MPI wrappers
set tmp1 default_str

set tmp $tmp1
catch {set tmp [ exec which mpiicc ]}
setenv CC $tmp
set tmp $tmp1
catch {set tmp [ exec which mpiicpc ]}
setenv CXX $tmp
set tmp $tmp1
catch {set tmp [ exec which mpiifort ]}
setenv F90 $tmp
setenv FC  $tmp
set tmp $tmp1
catch {set tmp [ exec which mpirun ]}
setenv MPIEXEC $tmp

```

The result in python is that `FC` `CC` and `CXX` all show as `default_str`.  Everything else about the modules work.  

Instead of this method, could we have `jedi-impi` module set these variables?

**Comment by mmiesch** [comment]
Are you running on sles12?  There intel 17 is not available - only intel 19, available with the apps/jedi/intel-impi/20.0.0.166 module.  That's the one that you should be using.

**Comment by rickgrubin** [comment]
I'd just logged in to discover and ssh'd to 'discover-sles12' (must first
go to 'discover' and then ssh to 'discover-sles12'); see:

https://jointcenterforsatellitedataassimilation-jedi-docs.readthedocs-hosted.com/en/latest/developer/jedi_environment/modules.html#discover

On discover, *module avail* shows the intel17 modules; that's not the case
on discover-sles12, where module  compiler/intel/
19.1.0.166/jedi-impi/19.1.0.166 is available.

--
Rick Grubin
Joint Center for Satellite Data Assimilation
UCAR Foothills Lab / Building 4
3300 Mitchell Lane
Boulder, CO 80301
grubin@ucar.edu

**Comment by mmiesch** [comment]
Thanks @rickgrubin - yes, that is the intended behavior.  So, if you're seeing the intel17 modules, it probably means your on sles-11.  NCCS asked us to move all JEDI operations to sles-12 so those are the modules I intend to keep updated.

**Comment by mmiesch** [comment]
Oh - also - I should get rid of 19.1.0.166.  This points to the identical compiler/mpi directories as 20.0.0.166.  I built both with different compiler options when I was first experimenting around with discover but 20.0.0.166 is the one I intend to keep up to date.  For example, you should see that there is a v0.2 version of this available that has the new ecbuild, eckit, and odc updates.

**Comment by rickgrubin** [comment]
On discover-sles12, 'module avail' shows:

-- /discover/nobackup/mmiesch/modules/modulefiles --
   apps/jedi/gcc-7.3_openmpi-3.0.0-baselibs
   [...]
   apps/jedi/intel-17.0.7.259-baselibs
   apps/jedi/intel-17.0.7.259
   apps/jedi/intel17-impi
(L,D)    <----  (L)oaded, and (D)efault module
   [...]

intel17-impi is: intel-17.0.7.259

This is set in: .../jedi-rapids/scripts/module-load.sh, as:

case $(hostname) in
[...]
discover*|borg*)
    . /usr/share/modules/init/bash
    export OPT=/discover/nobackup/mmiesch/modules
    module use -a $OPT/modulefiles
    module load other/anaconda3 other/git
    module load apps/jedi/intel17-impi
    ;;

A quick try at:

 'module use /discover/swdev/jcsda/modules/modulefiles' && 'module load
apps/jedi/intel-impi/19.0.0.166'

did set the (L)oaded / (D)efault compiler as desired, I think:

-- /discover/swdev/jcsda/modules/modulefiles --
   apps/jedi/baselibs/intel-impi/19.1.0.166-v0.1
   apps/jedi/baselibs/intel-impi/19.1.0.166-v0.2
   apps/jedi/baselibs/intel-impi/19.1.0.166                         (D)
   apps/jedi/gnu-impi/9.2.0-v0.1
   apps/jedi/gnu-impi/9.2.0-v0.2
   apps/jedi/gnu-impi/9.2.0                                         (D)
   apps/jedi/intel-impi/default
   apps/jedi/intel-impi/19.1.0.166.                               (L)
   apps/jedi/intel-impi/20.0.0.166-v0.1
   apps/jedi/intel-impi/20.0.0.166-v0.2
   apps/jedi/intel-impi/20.0.0.166                                  (D)


As 19.0.0.166 and 20.0.0.166 are the same, this seems OK.

I commented out the module load for apps/jedi/intel17 and re-ran the
jedi-module-load.sh script; saw this message:

rgrubin@discover19:~/src/jedi-rapids$ ./scripts/jedi-module-load.sh
Lmod has detected the following error:  The following module(s) are unknown:
"other/git" "other/anaconda3"

Please check the spelling or version number. Also try "module spider ..."
It is also possible your cache file is out-of-date; it may help to try:
  $ module --ignore-cache load "other/git" "other/anaconda3"

Also make sure that all modulefiles written in TCL start with the string
#%Module

but proceeded onward and installed jedi-rapids; tried a build:

% jedi-rapids build --bundle rapids --clean

2020-05-14 18:57:39,555 [INFO]: Build [prefix: master]:
/discover/nobackup/rgrubin/jedirapids/prefix/master
2020-05-14 18:57:39,576 [INFO]: [RuntimeEnv] [INTERACTIVE] #procs:4
(limit:4)
2020-05-14 18:57:39,576 [INFO]: Build [prefix: master]: Using up to: 4 procs
2020-05-14 18:57:39,617 [INFO]: [LMod] Use:
/discover/nobackup/mmiesch/modules/modulefiles

Due to MODULEPATH changes, the following have been reloaded:
  1) core/boost-headers/1.68.0     2) core/eigen/3.3.5

The following have been reloaded with a version change:
  1) core/git-lfs/2.10.0 => core/git-lfs/2.9.0


============= Build context ==================
name: master
build_app: bundle
bundle: rapids
branch: develop
build_type: RelWithDebInfo
modules: intel17-impi       <-- oops
debug: false
warn: false
ecbuildflags: []
cmakeflags: []
fflags: []
cflags: []
cxxflags: []
ldflags: []
env: {}
source_path: /discover/nobackup/rgrubin/jedirapids/build/bundle/master
bundle_url: https://github.com/jcsda/rapids-bundle
build_path: /discover/nobackup/rgrubin/jedirapids/build/bundle/master/_build
================================================================================

2020-05-14 18:57:40,300 [INFO]: [LMod] Load system environment: intel17-impi
2020-05-14 18:57:40,300 [INFO]: [LMod] Load modules: apps/jedi/intel17-impi
Lmod has detected the following error: The following module(s) are unknown:
"lib/mkl-17.0.7.259" "comp/intel-17.0.7.259" "mpi/impi-17.0.7.259"
"other/anaconda3/2019.03" "other/wget" "other/git" "other/comp/gcc-6.3"
"other/cmake"
[...]

and the build fails. Somehow / somewhere I'm not unloading the intel17
module(s) correctly.

So two questions:

1. Mark M.: should there be an intel-19... module directory in $OPT above,
or should Mark O. run 'module use ...' && 'module load ...' from elsewhere?
2. How do I properly remove the intel17 module from the jedi-rapids build
context?

--
Rick Grubin
Joint Center for Satellite Data Assimilation
UCAR Foothills Lab / Building 4
3300 Mitchell Lane
Boulder, CO 80301
grubin@ucar.edu

**Comment by mmiesch** [comment]
@rickgrubin - as far as I am concerned, users should not be using the intel 17 modules at all.  I think the problem (one of them, anyway) is that your `$OPT` variable is set incorrectly.   That's the answer to question 1.  The module on `/discover/nobackup` are the old sles-11 ones that have been deprecated - they may still work but will not be maintained going forward.  `$OPT` should now point to `/discover/swdev`.  So, that's also the answer to your question 2, I think, unless the jedi-rapids script makes explicit reference to sles-11 modules like `other/anaconda`.  Those will have to be removed too.

**Comment by rickgrubin** [comment]
If I take these steps:


   1. clone jedi-rapids repository
   2. edit appropriate script to comment out $OPT
   3. unload / load modules so as to point at /discover/swdev


and then start the configure process, I still see a build context that
shows:

modules: intel17-impi       <-- oops

So I was rather sure I knew the answer to question #1, thanks for
confirming.

I'll have to better understand how jedi-rapids sets its build context in
order to better answer question #2 and use appropriate modules.

I'm rusty with lmod; removing the .lmod.d/.cache directory can't hurt prior
to starting from scratch (as it were), true?

--
Rick Grubin
Joint Center for Satellite Data Assimilation
UCAR Foothills Lab / Building 4
3300 Mitchell Lane
Boulder, CO 80301
grubin@ucar.edu

**Comment by mmiesch** [comment]
true - go ahead and wipe your `.lmod.d/.cache` with abandon - it won't hurt anything.

**Comment by mmiesch** [comment]
And - no need to just comment out `$OPT` - it should be set to the `/discover/swdev` directory described in the docs - if my memory serves, I think it's `/discover/swdev/jcsda/modules`

**Comment by markjolah** [comment]
I'm going to close this as we're going to move to SLES12 ( jedi-rapids #73 ), and most of the mis-understanding was the older JEDI-Rapids was still setup for the the SLES11 modules that should now be fixed with #73.
